### PR TITLE
Torchhub: rewrite commit hash check to avoid using unnecessary GitHub API credits

### PR DESCRIPTION
**Pull request by NicolasHug**
This PR adds more detailed error messages to torchhub if the commit hash validation goes wrong, providing suggestions to the users on how to resolve the issue.

It also documents why such validation is important.

EDIT: it also avoids validatating some stuff when we know "stuff" isn't a commit since there's no risk in this case

CC @malfet @mthrok 

cc @nairbv @NicolasHug

### CODE REVIEW

File: torch/hub.py
Comment (NicolasHug) [inline torch/hub.py:142]: For reviewers: I didn't change this block at all apart from the error message at the end. I just put it inside a try/except statement so everything is indented to the right

File: torch/hub.py
Comment (mthrok) [inline torch/hub.py:126]: > weird

Although this gave me a moment of chuckle, I am afraid this might be taken as offensive. I imagine there was some technical reasons behind this for the sake of scalability or something.

File: torch/hub.py
Comment (mthrok) [inline torch/hub.py:133]: nit: I personally found this story interesting and informative on the perspective of how GitHub works behind the scene, but this particular instance was a troll made around a serious law suite, so I would not put it in here, just to be safe. I think the goal of communicating the potential security issue is achieved even without it.

File: torch/hub.py
Comment (mthrok) [reply torch/hub.py:142]: I think this is a neat update and it surely improves the user experience. However the cyclomatic complexity of this whole `try~except~raise` is very high. So I suggest to extract the logic around `while True` block, as a follow up.

**Review by mthrok — APPROVED** [review]
Stamp

File: torch/hub.py
Comment (NicolasHug) [reply torch/hub.py:141]: thanks, that makes sense. I extracted the "find branch" logic into a separate subfunction, hopefully the code looks simpler now

File: torch/hub.py
Comment (malfet) [inline torch/hub.py:139]: Aren't pull request a branch in github parlance? I.e. creating a pull request and passing `pull/12345` as branch name will bypass the check and has a potential of loading random code, isn't it?

Comment (NicolasHug) [reply torch/hub.py:139]: oh, thanks for looking at this @malfet , I actually wanted to check with you before merging it to make sure it wasn't introducing a breach.

As far as I can tell the culprit here is the download URL `f"https://github.com/{repo_owner}/{repo_name}/archive/{branch}.zip"`

but even if PR are considered to be a branch, the URL wouldn't exist. For example:

```py
torch.hub.load('pytorch/vision:pull/123', 'vgg11', skip_validation=True)
```

would try to download https://github.com/pytorch/vision/archive/pull/123.zip which leads to a 404

File: torch/hub.py
Comment (malfet) [inline torch/hub.py:167]: I somewhat hate infinite loops in the code. Can we perhaps limit the search to say 1000 or 5000 branches/tags? (For all practical purposes it should be enough)
```suggestion
            # Limit the search to 5000 branches/tags
            for page in range(1, 51):
                page += 1
                url = f'{url_prefix}?per_page=100&page={page}'
                response = json.loads(_read_url(Request(url, headers=headers)))
                # Empty response means no more data to process
                if not response:
                    break
                for br in response:
                    if br['name'] == branch or br['commit']['sha'].startswith(branch):
                        return True
            raise RuntimeError("How many branches must a script traverse\nBefore it exhausted the search")
```

**Comment by NicolasHug** [comment]
Errr so apparently I messed up a merge and it automatically requested reviews from a lot of people. Sorry for the noise everyone!


image

File: torch/hub.py
Comment (malfet) [reply torch/hub.py:139]: Thank you for the explanation. Makes sense then.

File: torch/hub.py
Comment (malfet) [inline torch/hub.py:160]: I would've printed some warning here...

File: torch/hub.py
Comment (malfet) [inline torch/hub.py:136]: `string.hexdigits=='0123456789abcdefABCDEF'` i.e. lowering it and then comparing it lowered branch name might be an overkill. 
```suggestion
    if any(c not in string.hexdigits for c in branch):
```

File: torch/hub.py
Comment (malfet) [inline torch/hub.py:139]: Also, I wonder if this would be a bit more Pythonic way of doing the same check
```suggestion
    try:
       int(branch,16)
    except ValueError:
        # The above issue only exists for commit hashes. So if we know that ``branch``
        # isn't a commit hash, we can return early. This should avoid many GitHub API calls
        return
```

**Review by malfet — APPROVED** [review]

File: torch/hub.py
Comment (mthrok) [inline torch/hub.py:188]: maybe you want to reset the exception context.

```suggestion
            ) from None
```

File: torch/hub.py
Comment (mthrok) [inline torch/hub.py:145]: Is there a strong reason why this has to be inner function?

I do not oppose much but I get the impression that if we extract these, the `_validate_branch` function only expresses the logical flow, and it's very readable

Comment (NicolasHug) [reply torch/hub.py]: I agree extracting it would make it simpler. Done!

File: torch/hub.py
Comment (NicolasHug) [reply torch/hub.py]: Good point, actually I think we should do `from e` to properly preserve the whole stack (it provides more details for the user while still keeping the stack clean)

**Comment by NicolasHug** [comment]
`test_load_commit_from_forked_repo` is passing on a few jobs, I'll merge (after fixing the linting).

Thanks a lot for the reviews @mthrok and @malfet !!

**Comment by mruberry** [comment]
Unlanding. This appears to have broken the ROCm job:

```
18:59:27 ======================================================================
18:59:27 FAIL [0.087s]: test_load_commit_from_forked_repo (__main__.TestHub)
18:59:27 ----------------------------------------------------------------------
18:59:27 Traceback (most recent call last):
18:59:27   File "/opt/conda/lib/python3.6/site-packages/torch/hub.py", line 178, in _validate_branch
18:59:27     branch_found = _branch_belongs_to_repo(repo_owner, repo_name, branch)
18:59:27   File "/opt/conda/lib/python3.6/site-packages/torch/hub.py", line 153, in _branch_belongs_to_repo
18:59:27     return any(find_in_refs(ref_kind) for ref_kind in ('branches', 'tags'))
18:59:27 urllib.error.HTTPError: HTTP Error 403: rate limit exceeded
18:59:27 
18:59:27 The above exception was the direct cause of the following exception:
18:59:27 
18:59:27 ValueError: Torchhub was unable to verify that 4e2c216 is indeed part of the pytorch/vision repo, as it received a GitHub API rate limit error. Consider setting the GITHUB_TOKEN env variable to a GitHub token that has more API credits. If you're absolutely sure about what you are doing, you can bypass this security check by setting skip_validation=True.
18:59:27 
18:59:27 During handling of the above exception, another exception occurred:
18:59:27 
18:59:27 Traceback (most recent call last):
18:59:27   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_utils.py", line 2249, in f_retry
18:59:27     return f(*args, **kwargs)
18:59:27   File "test_utils.py", line 706, in test_load_commit_from_forked_repo
18:59:27     torch.hub.load('pytorch/vision:4e2c216', 'resnet18', force_reload=True)
18:59:27 AssertionError: "Torchhub tried to look for 4e2c216" does not match "Torchhub was unable to verify that 4e2c216 is indeed part of the pytorch/vision repo, as it received a GitHub API rate limit error. Consider setting the GITHUB_TOKEN env variable to a GitHub token that has more API credits. If you're absolutely sure about what you are doing, you can bypass this security check by setting skip_validation=True."
```

**Pull request closed without merging**
